### DISCUSSION TITLE
Companions and other Moor objects are cluttering up memory

### DISCUSSION THREAD

**Discussion by lisaaerioh** [simolus3/drift#1911] [Q&A]
Hi, I'm using Moor for an app that collects sensor data and inserts the data into a database. A lot of sensor data can be entered into the database at a time, and for each of those insertions, I've found the Companions (and other Moor objects) are being created and are not immediately removed. With a lot of insertions over time, this causes the app to use a lot of memory and affects the performance & stability of the app. I'm currently opening up the database in a background thread according to the docs. My latest idea is to open up the database in the main thread for most reading/writing operations, but open/close the database in a background thread for writing a large amount of sensor data. Is something like this possible or would it corrupt the database to have it opened in multiple threads?

Is this expected behavior or am I doing something incorrectly, and is there an easy way to prevent this memory clutter?

Thank you!

**Comment by simolus3** [comment]
Thanks for starting this discussion!

> A lot of sensor data can be entered into the database at a time, and for each of those insertions, I've found the Companions (and other Moor objects) are being created and are not immediately removed.

I'm not sure what you mean with "not immediately removed". They are designed to be very short-lived objects, so they should be picked up during the next GC run. If drift is causing a memory leak, then obviously this needs to be fixed, but I don't see how anything could keep a reference to companions? If you could share some insights on how you uncovered this memory usage, I'm happy to try reproducing it myself and work on fixing it.

> My latest idea is to open up the database in the main thread for most reading/writing operations, but open/close the database in a background thread for writing a large amount of sensor data.

This should be fine, even without the isolates feature that moor/drift provides to you. sqlite3 can handle concurrent connections to the same file, but you may run into issues about the two threads locking each other. Enabling [write-ahead logging](https://sqlite.org/wal.html) in sqlite helps with that.

> Is this expected behavior or am I doing something incorrectly, and is there an easy way to prevent this memory clutter?

I'm not sure if that's on you or due to a bug in drift, but companions you use for inserts are definitely not supposed to stick around. Again, I'd love to take a better look if you can share more details on this.

**Reply by lisaaerioh** [reply]
Hi,

I discovered this using the Memory Page in Flutter Dev Tools. I will continue to do some testing, but it seems when entering into the database, objects such as Companions are created, and even after the garbage collector removes objects some of these objects are still there and build up over time. It does seem like they get removed after a while, but not before they've built up and caused memory pressure. Sorry the details are vague, I'm still testing the behavior of Moor & garbage collection, but will get back to you with more details.

Meanwhile, the way I've been using Moor is by opening the database in a background thread according to the docs and using a Riverpod provider around the database singleton (using the DatabaseConnection) to access the database throughout the app. Let me know if this is enough information for you to try to recreate the issue or if you need any other details.

I've also tried opening the database in the main thread using the original constructor and also opening the database in a background thread using the Moor isolate feature for write operations, and I'm getting a warning that the database was opened in two places and can cause corruption. I'm afraid I don't have much expertise with SQL. Am I correct in understanding that as long as we take the necessary precautions to prevent corruption, we can go ahead and use the database like this? Are there any other things besides enabling write-ahead logging that I should be aware of in order to prevent corruption? 

Thank you for all the help!
